### PR TITLE
Conditional Template

### LINKED ISSUE #2741: Enable optional fields to show/hide in UI
As templates become more advanced, we need to support a more dynamic form template, implementing support for `allOf`, `if/else` in the template json. 

Some UI-specific changes will need to be made as the jrsf library is pretty fussy about rendering when there are null fields in the template.

### PR DESCRIPTION
**Pull request by damoodamoo**
# Resolves #2741 

- Implements the `allOf` feature of JSON schema to allow conditional fields + conditional schema validation
- "auto_create" string has gone in favour of an enum (dropdown), which shows appropriate follow-up fields
- all workspace templates updated to use conditional schema to streamline the form
- Airlock review VM configuration - nested properties, conditionally required
- UI code to massage the data payload sent in for create or update of a resource. 

image

### CODE REVIEW

File: ui/app/src/models/resourceTemplate.ts [hunk 1]
```diff
@@ -1,32 +1,64 @@
...
+  allOf?: Array<any>,
+  system_properties: any,
+  actions: Array<TemplateAction>,
+  customActions: Array<TemplateAction>,
+  required: Array<string>,
+  uiSchema: any,
+  pipeline: any
 }
 
+export const sanitiseTemplateForRJSF = (template: ResourceTemplate) => {
```
Comment (tanya-borisova) [inline ui/app/src/models/resourceTemplate.ts:22]: FWIW this doesn't look that horrible to me although it is obviously a hack. Perhaps worth a second pair of eyes though

**Review by tanya-borisova — APPROVED** [review]
LGTM in the API app domain but I think it would need another approval for the UI part.

Exciting changes! 🎉

File: api_app/tests_ma/test_api/test_routes/test_airlock.py [hunk 2]
```diff
@@ -105,13 +105,13 @@ def sample_airlock_review_config() -> dict:
     return {
         "airlock_review_config": {
             "import": {
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply api_app/tests_ma/test_api/test_routes/test_airlock.py:107]: not in the end, I was able to keep them nested

File: api_app/schemas/azuread.json [hunk 3]
```diff
@@ -5,19 +5,8 @@
...
-            "title": "Application (Client) ID",
-            "description": "The AAD Application Registration ID for the workspace. Use 'auto_create' if you wish TRE to create this."
-        },
-        "client_secret": {
-          "type": "string",
-          "title": "Application (Client) Secret",
-          "description": "The AAD Application Registration secret for the workspace. Leave blank if using `auto_create` above. This value will be stored in the Workspace Key Vault.",
-          "sensitive": true
-      }
+
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply api_app/schemas/azuread.json:10]: well not really - BUT - there's no harm leaving it here for now. There's a chance we might revisit all these static templates in time.

File: CHANGELOG.md [hunk 4]
```diff
@@ -23,6 +23,7 @@ ENHANCEMENTS:
 * Show error message when Review VMs are not configured in the current workspace
 * CLI: Add missing endpoints and minor bug fixes (#2784)
 * Added optional parameter to allow a client to retrieve a template by name and version ([#2802](https://github.com/microsoft/AzureTRE/pull/2802))
+* Added support for `allOf` usage in Resource Templates - both across the API and the UI. This allows a template author to specify certain fields as being conditionally present / conditionally required, and means we can tidy up some of the resource creation forms substantially. ([#2795](https://github.com/microsoft/AzureTRE/pull/2795))
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply CHANGELOG.md]: yup, good call - have added

File: api_app/api/routes/airlock.py [hunk 5]
```diff
@@ -130,13 +130,15 @@ async def create_review_user_resource(
...
+            workspace_id = config["import_vm_workspace_id"]
+            workspace_service_id = config["import_vm_workspace_service_id"]
+            user_resource_template_name = config["import_vm_user_resource_template_name"]
         else:
             assert airlock_request.type == AirlockRequestType.Export
             config = workspace.properties["airlock_review_config"]["export"]
             workspace_id = workspace.id
-        workspace_service_id = config["workspace_service_id"]
-        user_resource_template_name = config["user_resource_template_name"]
+            workspace_service_id = config["export_vm_orkspace_service_id"]
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply api_app/api/routes/airlock.py]: good spot!

File: ui/app/src/components/shared/create-update-resource/ResourceForm.tsx [hunk 6]
```diff
@@ -30,20 +31,21 @@ export const ResourceForm: React.FunctionComponent<ResourceFormProps> = (props:
     const getFullTemplate = async () => {
       try {
         // Get the full resource template containing the required parameters
-        const templateResponse = await apiCall(props.updateResource ? `${props.templatePath}?is_update=true` : props.templatePath, HttpMethod.Get);
+        const templateResponse = (await apiCall(props.updateResource ? `${props.templatePath}?is_update=true&version=${props.updateResource.templateVersion}` : props.templatePath, HttpMethod.Get)) as ResourceTemplate;
 
         // if it's an update, populate the form with the props that are available in the template
         if (props.updateResource) {
           let d: any = {};
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply ui/app/src/components/shared/create-update-resource/ResourceForm.tsx:38]: good spot, i'd reworked that and left it there!

File: ui/app/src/components/shared/create-update-resource/ResourceForm.tsx [hunk 7]
```diff
@@ -56,23 +58,69 @@ export const ResourceForm: React.FunctionComponent<ResourceFormProps> = (props:
...
+          recurseTemplate(templateFragment[key]);
+        }
+      })
+    }
+
+    recurseTemplate(template);
+
+    // strip out the properties that are not on-screen at time of sending.
+    // if the properties aren't on the screen it means they're in conditional parts of the template and shouldn't be being sent
+    let onScreenPropLabelElements = document.getElementsByClassName('rjsf')[0].getElementsByClassName('ms-Label');
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply ui/app/src/components/shared/create-update-resource/ResourceForm.tsx]: faaaaaaaaaaanx!

File: ui/app/src/components/shared/create-update-resource/ResourceForm.tsx [hunk 8]
```diff
@@ -30,20 +31,21 @@ export const ResourceForm: React.FunctionComponent<ResourceFormProps> = (props:
     const getFullTemplate = async () => {
       try {
         // Get the full resource template containing the required parameters
-        const templateResponse = await apiCall(props.updateResource ? `${props.templatePath}?is_update=true` : props.templatePath, HttpMethod.Get);
+        const templateResponse = (await apiCall(props.updateResource ? `${props.templatePath}?is_update=true&version=${props.updateResource.templateVersion}` : props.templatePath, HttpMethod.Get)) as ResourceTemplate;
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply ui/app/src/components/shared/create-update-resource/ResourceForm.tsx:34]: yup, broken it up..

**Review by jjgriff93 — APPROVED** [review]
Perfect, looks great!

**Pull request merged**